File: release/ATS-Resume-Studio/ATS-Resume-Studio-App/serve.py
#!/usr/bin/env python3
"""
ATS Resume Studio - Local Zero-Dependency Server
100% In-Browser & Private. No external APIs or network calls.
"""
import http.server
import socketserver
import socket
import os
import sys
import time
import threading
import subprocess

# 1. Base application directory strictly on the location of serve.py (never relies on cwd)
DIRECTORY = os.path.dirname(os.path.abspath(__file__))

def find_open_port(start_port=3000, max_attempts=20):
    """Finds an available local port on 127.0.0.1 without conflicts."""
    for port in range(start_port, start_port + max_attempts):
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            try:
                s.bind(('127.0.0.1', port))
                return port
            except OSError:
                continue
    return start_port

class ResumeRequestHandler(http.server.SimpleHTTPRequestHandler):
    extensions_map = {
        **http.server.SimpleHTTPRequestHandler.extensions_map,
        '.mjs': 'application/javascript',
        '.js': 'application/javascript',
        '.wasm': 'application/wasm',
        '.gz': 'application/gzip',
        '.webp': 'image/webp',
    }

    def __init__(self, *args, **kwargs):
        super().__init__(*args, directory=DIRECTORY, **kwargs)

    def do_GET(self):
        # Clean route mapping for static export
        clean_path = self.path.split('?')[0].split('#')[0]
        if clean_path in ('', '/', '/app'):
            self.path = '/app/'
        elif clean_path == '/privacy':
            self.path = '/privacy/'
        elif clean_path == '/terms':
            self.path = '/terms/'
        return super().do_GET()

    def log_message(self, format, *args):
        # Keep terminal output clean
        pass

def wait_and_open_browser(url, port):
    """Verifies that the server is actively listening on 127.0.0.1 before opening browser."""
    connected = False
    for _ in range(40):
        try:
            with socket.create_connection(('127.0.0.1', port), timeout=0.2):
                connected = True
                break
        except OSError:
            time.sleep(0.05)

    if connected:
        try:
            if sys.platform == 'darwin':
                subprocess.Popen(['open', url], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
            elif sys.platform.startswith('win'):
                os.startfile(url)
            else:
                subprocess.Popen(['xdg-open', url], stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL)
        except Exception:
            pass

def run():
    # Ensure working directory matches script directory
    try:
        os.chdir(DIRECTORY)
    except Exception:
        pass

    port = find_open_port(3000)
    server_address = ('127.0.0.1', port)
    
    socketserver.TCPServer.allow_reuse_address = True
    try:
        with socketserver.TCPServer(server_address, ResumeRequestHandler) as httpd:
            url = f"http://127.0.0.1:{port}/app/"
            python_version = f"{sys.version_info.major}.{sys.version_info.minor}.{sys.version_info.micro}"

            print("============================================================")
            print("   ATS RESUME STUDIO — LOCAL SERVER RUNNING")
            print("============================================================")
            print(f"   Status:        Listening on 127.0.0.1:{port} (Active)")
            print(f"   Server URL:    {url}")
            print(f"   Engine:        Python {python_version}")
            print(f"   App Directory: {DIRECTORY}")
            print("------------------------------------------------------------")
            print("   Opening your web browser automatically...")
            print("   KEEP THIS TERMINAL WINDOW OPEN while using the app.")
            print("   Press Ctrl+C to shut down the server when finished.")
            print("============================================================")
            sys.stdout.flush()

            # Trigger browser opening strictly after listener is active
            browser_thread = threading.Thread(target=wait_and_open_browser, args=(url, port), daemon=True)
            browser_thread.start()

            httpd.serve_forever()
    except KeyboardInterrupt:
        print("\n\nATS Resume Studio server stopped. Have a wonderful day!")
        sys.exit(0)
    except Exception as e:
        print(f"\n❌ Local server error: {e}", file=sys.stderr)
        sys.exit(1)

if __name__ == '__main__':
    run()

File: release/ATS-Resume-Studio/ATS-Resume-Studio-App/serve.js
#!/usr/bin/env node
/**
 * ATS Resume Studio - Local Zero-Dependency Server (Node.js Fallback)
 * 100% In-Browser & Private. No external APIs or network calls.
 */
const http = require('http');
const fs = require('fs');
const path = require('path');
const net = require('net');
const { exec } = require('child_process');

// 1. Base application directory strictly on __dirname (never relies on cwd)
const DIRECTORY = __dirname;
let currentPort = 3000;

const MIME_TYPES = {
  '.html': 'text/html; charset=utf-8',
  '.css': 'text/css; charset=utf-8',
  '.js': 'application/javascript; charset=utf-8',
  '.mjs': 'application/javascript; charset=utf-8',
  '.wasm': 'application/wasm',
  '.gz': 'application/gzip',
  '.json': 'application/json; charset=utf-8',
  '.svg': 'image/svg+xml',
  '.png': 'image/png',
  '.jpg': 'image/jpeg',
  '.jpeg': 'image/jpeg',
  '.webp': 'image/webp',
  '.ico': 'image/x-icon',
  '.txt': 'text/plain; charset=utf-8',
};

function verifyAndOpenBrowser(url, port) {
  let attempts = 0;
  function checkSocket() {
    attempts++;
    const client = net.createConnection({ port, host: '127.0.0.1' }, () => {
      client.end();
      const openCmd =
        process.platform === 'darwin'
          ? `open "${url}"`
          : process.platform === 'win32'
          ? `start "" "${url}"`
          : `xdg-open "${url}"`;
      exec(openCmd);
    });
    client.on('error', () => {
      if (attempts < 40) {
        setTimeout(checkSocket, 50);
      }
    });
  }
  checkSocket();
}

function startServer(port) {
  const server = http.createServer((req, res) => {
    let reqPath = decodeURI(req.url.split('?')[0]);
    if (reqPath === '' || reqPath === '/' || reqPath === '/app') reqPath = '/app/';
    else if (reqPath === '/privacy') reqPath = '/privacy/';
    else if (reqPath === '/terms') reqPath = '/terms/';

    if (reqPath.endsWith('/')) reqPath += 'index.html';

    const filePath = path.join(DIRECTORY, reqPath);
    fs.stat(filePath, (err, stats) => {
      if (!err && stats.isFile()) {
        const ext = path.extname(filePath).toLowerCase();
        res.writeHead(200, { 'Content-Type': MIME_TYPES[ext] || 'application/octet-stream' });
        fs.createReadStream(filePath).pipe(res);
      } else {
        res.writeHead(404, { 'Content-Type': 'text/plain' });
        res.end('404 Not Found');
      }
    });
  });

  server.on('error', (err) => {
    if (err.code === 'EADDRINUSE') {
      startServer(port + 1);
    } else {
      console.error(`❌ Server error: ${err.message}`);
      process.exit(1);
    }
  });

  server.listen(port, '127.0.0.1', () => {
    const url = `http://127.0.0.1:${port}/app/`;
    console.log('============================================================');
    console.log('   ATS RESUME STUDIO — LOCAL SERVER RUNNING');
    console.log('============================================================');
    console.log(`   Status:        Listening on 127.0.0.1:${port} (Active)`);
    console.log(`   Server URL:    ${url}`);
    console.log(`   Engine:        Node.js (${process.version})`);
    console.log(`   App Directory: ${DIRECTORY}`);
    console.log('------------------------------------------------------------');
    console.log('   Opening your web browser automatically...');
    console.log('   KEEP THIS TERMINAL WINDOW OPEN while using the app.');
    console.log('   Press Ctrl+C to shut down the server when finished.');
    console.log('============================================================');

    verifyAndOpenBrowser(url, port);
  });
}

process.on('SIGINT', () => {
  console.log('\n\nATS Resume Studio server stopped. Have a wonderful day!');
  process.exit(0);
});

startServer(currentPort);
